inject_file_template NUL bytes false positive in launch-agent.sh

## Description

`launch-agent.sh:1554` rejects valid `inject_file_template` with:
```
[ERROR] inject_file_template for CLAUDE_OAUTH contains NUL bytes
```

The template is clean (verified via isolated base64 decode), and the secret value from `security find-generic-password -w` also has no NUL bytes. The agent exits with code 1 and the Slack bot reports `container_crash`.

## Environment
- Kapsis 2.19.0 (homebrew, aarch64-apple-darwin)
- bash 5.3.9
- macOS Darwin 25.3.0

## Config (agent YAML)
```yaml
CLAUDE_OAUTH:
  service: "claude-oauth-token"
  account: "aviad.s"
  inject_to_file: "~/.claude/.credentials.json"
  inject_file_template: |
    {"claudeAiOauth": {"accessToken": "{{VALUE}}"}}
  mode: "0600"
```

## Analysis

The NUL check at line 1554:
```bash
if [[ "$template_raw" == *$'\0'* ]]; then
    log_error "inject_file_template for $var_name contains NUL bytes"
    exit 1
fi
```

In isolated testing, `base64 -d` of the yq-encoded template decodes cleanly with no NUL bytes. The check does not trigger. However, in the actual launch-agent context (within the `while IFS='|' read` loop with all env vars being resolved), it triggers consistently on retry.

Possible causes:
1. A prior variable in the `while` loop corrupts shell state
2. Interaction between `query_secret_store_with_fallbacks` and the `base64 -d` command substitution
3. The `CREDENTIAL_FILES` or `SECRET_ENV_VARS` array accumulation affects the subshell environment

## Kapsis log
```
[01:49:43] [INFO] Loaded AIRFLOW_PASS from secret store (service: ldap-password)
[01:49:43] [INFO] Loaded BITBUCKET_TOKEN from secret store (service: taboola-bitbucket)
[01:49:43] [INFO] Loaded CLAUDE_OAUTH from secret store (service: claude-oauth-token)
[01:49:44] [ERROR] inject_file_template for CLAUDE_OAUTH contains NUL bytes
```

Note: CLAUDE_OAUTH is loaded successfully (line 3), but the template validation fails immediately after.

## Impact

Blocks all agent launches using `inject_file_template` with this config. Retry produces the same failure.

## Workaround

None known — the `inject_file_template` feature is the only way to pass structured credential files (Issue #241).

--- a/scripts/launch-agent.sh
+++ b/scripts/launch-agent.sh
@@ -1558,8 +1558,13 @@ generate_env_vars() {
 
                     # Skip empty templates (field is base64("") when not specified)
                     if [[ -n "$template_raw" ]]; then
-                        # Reject NUL bytes (would truncate shell strings)
-                        if [[ "$template_raw" == *$'\0'* ]]; then
+                        # Reject NUL bytes (would corrupt template substitution)
+                        # Bash variables silently strip NUL, so [[ == *$'\0'* ]]
+                        # always matches (Bug #251). Check raw byte stream instead.
+                        local _raw_count _clean_count
+                        _raw_count=$(printf '%s' "$inject_file_template_b64" | base64 -d 2>/dev/null | wc -c)
+                        _clean_count=$(printf '%s' "$inject_file_template_b64" | base64 -d 2>/dev/null | tr -d '\0' | wc -c)
+                        if (( _raw_count != _clean_count )); then
                             log_error "inject_file_template for $var_name contains NUL bytes"
                             exit 1
                         fi

--- a/scripts/launch-agent.sh
+++ b/scripts/launch-agent.sh
@@ -1560,11 +1560,12 @@ generate_env_vars() {
                     if [[ -n "$template_raw" ]]; then
                         # Reject NUL bytes (would corrupt template substitution)
                         # Bash variables silently strip NUL, so [[ == *$'\0'* ]]
-                        # always matches (Bug #251). Check raw byte stream instead.
-                        local _raw_count _clean_count
-                        _raw_count=$(printf '%s' "$inject_file_template_b64" | base64 -d 2>/dev/null | wc -c)
-                        _clean_count=$(printf '%s' "$inject_file_template_b64" | base64 -d 2>/dev/null | tr -d '\0' | wc -c)
-                        if (( _raw_count != _clean_count )); then
+                        # always matches (Bug #251). Compare raw byte count from
+                        # the base64 stream against NUL-stripped count instead.
+                        local raw_byte_count clean_byte_count
+                        raw_byte_count=$(printf '%s' "$inject_file_template_b64" | base64 -d 2>/dev/null | wc -c)
+                        clean_byte_count=$(printf '%s' "$inject_file_template_b64" | base64 -d 2>/dev/null | tr -d '\0' | wc -c)
+                        if (( raw_byte_count != clean_byte_count )); then
                             log_error "inject_file_template for $var_name contains NUL bytes"
                             exit 1
                         fi

--- a/tests/test-credential-file-template.sh
+++ b/tests/test-credential-file-template.sh
@@ -359,6 +359,42 @@ test_template_validation_in_launch_script() {
         "launch-agent.sh should cap marker count"
 }
 
+test_nul_check_no_false_positive() {
+    log_test "Testing NUL byte check does not false-positive on valid template (Bug #251)"
+
+    # The old check [[ "$var" == *$'\0'* ]] always matches because bash
+    # variables cannot hold NUL bytes — $'\0' in a pattern is empty string,
+    # making *$'\0'* equivalent to * (matches everything).
+    # The new check compares raw byte counts via pipe.
+
+    # Valid template — should pass NUL check
+    local template='{"token": "{{VALUE}}"}'
+    local template_b64
+    template_b64=$(printf '%s' "$template" | base64)
+
+    local _raw_count _clean_count
+    _raw_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | wc -c)
+    _clean_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | tr -d '\0' | wc -c)
+
+    assert_equals "$_raw_count" "$_clean_count" \
+        "Valid template should have equal raw and clean byte counts"
+}
+
+test_nul_check_detects_actual_nul() {
+    log_test "Testing NUL byte check detects actual NUL bytes in template"
+
+    # Create a template WITH an embedded NUL byte, base64-encode it
+    local template_b64
+    template_b64=$(printf 'hello\0world' | base64)
+
+    local _raw_count _clean_count
+    _raw_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | wc -c)
+    _clean_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | tr -d '\0' | wc -c)
+
+    assert_true "(( _raw_count != _clean_count ))" \
+        "Template with NUL should have different raw vs clean byte counts"
+}
+
 test_entrypoint_has_security_invariant() {
     log_test "Testing entrypoint.sh has SECURITY INVARIANT comment"
 
@@ -400,6 +436,8 @@ main() {
 
     # Launch-agent validation tests (file content checks)
     run_test test_template_validation_in_launch_script
+    run_test test_nul_check_no_false_positive
+    run_test test_nul_check_detects_actual_nul
     run_test test_entrypoint_has_security_invariant
     run_test test_entrypoint_unsets_credential_files
 

--- a/tests/test-credential-file-template.sh
+++ b/tests/test-credential-file-template.sh
@@ -372,11 +372,11 @@ test_nul_check_no_false_positive() {
     local template_b64
     template_b64=$(printf '%s' "$template" | base64)
 
-    local _raw_count _clean_count
-    _raw_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | wc -c)
-    _clean_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | tr -d '\0' | wc -c)
+    local raw_byte_count clean_byte_count
+    raw_byte_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | wc -c)
+    clean_byte_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | tr -d '\0' | wc -c)
 
-    assert_equals "$_raw_count" "$_clean_count" \
+    assert_equals "$raw_byte_count" "$clean_byte_count" \
         "Valid template should have equal raw and clean byte counts"
 }
 
@@ -387,11 +387,11 @@ test_nul_check_detects_actual_nul() {
     local template_b64
     template_b64=$(printf 'hello\0world' | base64)
 
-    local _raw_count _clean_count
-    _raw_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | wc -c)
-    _clean_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | tr -d '\0' | wc -c)
+    local raw_byte_count clean_byte_count
+    raw_byte_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | wc -c)
+    clean_byte_count=$(printf '%s' "$template_b64" | base64 -d 2>/dev/null | tr -d '\0' | wc -c)
 
-    assert_true "(( _raw_count != _clean_count ))" \
+    assert_true "(( raw_byte_count != clean_byte_count ))" \
         "Template with NUL should have different raw vs clean byte counts"
 }
 

